### PR TITLE
Prevent insertion of nil into foundation collections

### LINKED ISSUE #1607: Mutable collections in foundation must not allow nil objects to be inserted
I think it's critical issue.

Example of code:
```
NSMutableDictionary *dictionary = [NSMutableDictionary dictionary];
dictionary[@"key1"] = @"string";
dictionary[@"key2"] = nil;
dictionary[@"key3"] = NULL;
dictionary[@"key4"] = [NSNull null];
```

On iOS dictionary has only **key1** and **key4**, print in debug:
```(lldb) po dictionary
{
    key1 = string;
    key5 = "<null>";
}
```

logged on WinObjC:
```
{
    key2 = (null)
    key3 = (null)
    key1 = string
    key5 = <null>
}
```

---
But need to remember that on iOS version of Foundation, next line cause exception:
`[dictionary setObject:nil forKey:@"key5"];`

Exception log:
```
2016-12-22 11:59:42.631 Test[44238:4517602] *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '*** setObjectForKey: object cannot be nil (key: key5)'
```

But on WinObjC, it will let you insert nil object.

### PR DESCRIPTION
**Pull request by aballway**
Now trying to insert into NSArray, NSDictionary, and NSAttributedString will throw an NSException

Fixes #1607

---
This change is [Reviewable](https://reviewable.io/reviews/microsoft/winobjc/1698)

### CODE REVIEW

File: tests/unittests/Foundation/NSDictionaryTests.mm [hunk 1]
```diff
@@ -130,3 +139,23 @@
...
+TEST(NSMutableDictionary, InsertingNilForKeyedSubscriptShouldRemoveValue) {
+    NSMutableDictionary* dict = [NSMutableDictionary dictionaryWithObject:@"world" forKey:@"hello"];
+    [dict setObject:nil forKeyedSubscript:@"hello"];
+    EXPECT_EQ(nil, dict[@"hello"]);
+    EXPECT_EQ(0, [dict count]);
+}
+
+TEST(NSMutableDictionary, InsertingNilWithLiteralSyntaxShouldRemoveValue) {
+    NSMutableDictionary* dict = [NSMutableDictionary dictionaryWithObject:@"world" forKey:@"hello"];
+    dict[@"hello"] = nil;
```
Comment (DHowett-MSFT) [inline tests/unittests/Foundation/NSDictionaryTests.mm]: "Literal syntax" only refers to `@{@"key": @"val"}`; you probably want Subscripting

File: tests/unittests/Foundation/NSDictionaryTests.mm [hunk 2]
```diff
@@ -130,3 +139,23 @@
 
     EXPECT_OBJCNE(input, output);
 }
+
+TEST(NSMutableDictionary, InsertingNilForKeyedSubscriptShouldRemoveValue) {
+    NSMutableDictionary* dict = [NSMutableDictionary dictionaryWithObject:@"world" forKey:@"hello"];
+    [dict setObject:nil forKeyedSubscript:@"hello"];
```
Comment (DHowett-MSFT) [inline tests/unittests/Foundation/NSDictionaryTests.mm]: for fun, add an EXPECT_NO_THROW around these.

File: Frameworks/Foundation/NSCFArray.mm [hunk 3]
```diff
@@ -106,6 +106,11 @@ - (void)removeLastObject {
 
 - (void)replaceObjectAtIndex:(NSUInteger)index withObject:(NSObject*)obj {
     BRIDGED_THROW_IF_IMMUTABLE(_CFArrayIsMutable, CFArrayRef);
+    if (obj == nil) {
```
Comment (rajsesh-msft) [inline Frameworks/Foundation/NSCFArray.mm]: This block of code is best turned into a macro COLLECTION_THROW_IF_NULL()

File: Frameworks/Foundation/NSCFArray.mm [hunk 4]
```diff
@@ -115,6 +120,12 @@ - (void)replaceObjectAtIndex:(NSUInteger)index withObject:(NSObject*)obj {
 
 - (void)insertObject:(NSObject*)objAddr atIndex:(NSUInteger)index {
     BRIDGED_THROW_IF_IMMUTABLE(_CFArrayIsMutable, CFArrayRef);
+    if (objAddr == nil) {
+        @throw [NSException exceptionWithName:NSInvalidArgumentException
+                                       reason:[NSString stringWithFormat:@"*** %@ object cannot be nil", NSStringFromSelector(_cmd)]
+                                     userInfo:nil];
+    }
```
Comment (msft-Jeyaram) [inline Frameworks/Foundation/NSCFArray.mm]: ++ repetitive code

File: Frameworks/Foundation/NSCFArray.mm [hunk 5]
```diff
@@ -125,6 +136,11 @@ - (void)removeAllObjects {
 
 - (void)addObject:(NSObject*)objAddr {
     BRIDGED_THROW_IF_IMMUTABLE(_CFArrayIsMutable, CFArrayRef);
+    if (objAddr == nil) {
+        @throw [NSException exceptionWithName:NSInvalidArgumentException
+                                       reason:[NSString stringWithFormat:@"*** %@ object cannot be nil", NSStringFromSelector(_cmd)]
+                                     userInfo:nil];
+    }
```
Comment (msft-Jeyaram) [inline Frameworks/Foundation/NSCFArray.mm]: use THROW_NS_IF_MSG
you don't have do the repetitive work.

File: Frameworks/Foundation/NSCFAttributedString.mm [hunk 6]
```diff
@@ -109,6 +113,12 @@ - (void)addAttribute:(NSString*)name value:(id)value range:(NSRange)range {
     BRIDGED_THROW_IF_IMMUTABLE(_CFAttributedStringIsMutable, CFAttributedStringRef);
     THROW_NS_IF_FALSE(E_BOUNDS, ((range.location + range.length) <= [self length]));
 
+    if (value == nil) {
+        @throw [NSException exceptionWithName:NSInvalidArgumentException
+                                       reason:[NSString stringWithFormat:@"*** %@ nil value", NSStringFromSelector(_cmd)]
+                                     userInfo:nil];
+    }
```
Comment (msft-Jeyaram) [inline Frameworks/Foundation/NSCFAttributedString.mm]: THROW_NS_IF_MSG

File: tests/unittests/Foundation/NSArrayTests.mm [hunk 7]
```diff
@@ -425,3 +425,13 @@ static unsigned long int objectIndexInArray(NSArray* array, int value, int start
 
     EXPECT_OBJCNE(input, output);
 }
+
+TEST(NSMutableArray, InsertingNilShouldThrow) {
+    NSMutableArray* arr = [NSMutableArray arrayWithObject:@"hello"];
+    EXPECT_ANY_THROW(arr[0] = nil);
+    EXPECT_ANY_THROW([arr insertObject:nil atIndex:1]);
+    EXPECT_ANY_THROW([arr addObject:nil]);
```
Comment (msft-Jeyaram) [inline tests/unittests/Foundation/NSArrayTests.mm]: check for the specific throw?
not sure if it's still an issue on arm.

Comment (DHowett-MSFT) [inline tests/unittests/Foundation/NSArrayTests.mm]: It absolutely is.
The only useful information we would get out of this is that it is a *NSInvalidArgumentException*. Unfortunately, that's not fine-grained enough to support what we care about.

"Any" exception is fine.

**Pull request merged**